### PR TITLE
feat(twap): init and export genesis

### PR DESCRIPTION
**Pull request by p0mvn**
Closes: #XXX

## What is the purpose of the change

Note to reviewers: This original large PR has been split into smaller to ease the review:
- https://github.com/osmosis-labs/osmosis/pull/2452
- https://github.com/osmosis-labs/osmosis/pull/2453

This change contains the logic for initializing and exporting TWAP genesis. The TWAP records are sorted by time during `InitGenesis` so that the "most recent" index is updated correctly.

## Documentation and Release Note

  - Does this pull request introduce a new feature or user-facing behavior changes? (yes / no)
  - Is a relevant changelog entry added to the `Unreleased` section in `CHANGELOG.md`? (yes / no)
  - How is the feature or change documented? (not applicable   /   specification (`x/<module>/spec/`)  /  [Osmosis docs repo](https://github.com/osmosis-labs/docs)   /   not documented)

### CODE REVIEW

File: x/twap/genesis.go [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+	twapRecords, err := k.getAllHistoricalPoolIndexedTWAPs(ctx)
+	if err != nil {
+		panic(err)
+	}
+
+	return &types.GenesisState{
+		Params: k.GetParams(ctx),
+		Twaps:  twapRecords,
+	}
+}
```
Comment (ValarDragon) [inline x/twap/genesis.go:34]: wdyt about moving this to keeper.go

File: x/twap/genesis.go [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+// state.
+func (k Keeper) InitGenesis(ctx sdk.Context, genState *types.GenesisState) {
+	if err := genState.Validate(); err != nil {
+		panic(err)
+	}
+
+	k.SetParams(ctx, genState.Params)
+
+	for _, twap := range genState.Twaps {
+		k.storeNewRecord(ctx, twap)
```
Comment (ValarDragon) [inline x/twap/genesis.go:19]: This requires the TWAPs to be sorted by time (or really for the last TWAP per pool to be called last)

We should either do the sorting here, or enforce the relevant property in validate

File: x/twap/twapmodule/module.go [hunk 3]
```diff
@@ -106,16 +106,22 @@ func (am AppModule) LegacyQuerierHandler(legacyQuerierCdc *codec.LegacyAmino) sd
 	}
 }
 
-// InitGenesis performs genesis initialization for the gamm module. It returns
+// InitGenesis performs genesis initialization for the twap module. It returns
```
Comment (p0mvn) [inline x/twap/twapmodule/module.go:109]:
```suggestion
// InitGenesis performs genesis initialization for the twap module.
```

File: x/twap/genesis.go (hunk 2 above)
Comment (p0mvn) [reply x/twap/genesis.go]: Fixed

File: x/twap/genesis.go (hunk 1 above)
Comment (p0mvn) [reply x/twap/genesis.go]: That makes sense, moved

**Comment by p0mvn** [comment]
Note to reviewers: I'm splitting this large PR into smaller to ease the review:
- https://github.com/osmosis-labs/osmosis/pull/2452
- https://github.com/osmosis-labs/osmosis/pull/2453

Please start from there. I will rebase this after the two are merged

File: x/twap/keeper.go [hunk 4]
```diff
@@ -40,3 +42,39 @@ func (k Keeper) SetParams(ctx sdk.Context, params types.Params) {
...
+// state.
+func (k Keeper) InitGenesis(ctx sdk.Context, genState *types.GenesisState) {
+	if err := genState.Validate(); err != nil {
+		panic(err)
+	}
+
+	k.SetParams(ctx, genState.Params)
+
+	// Most recent TWAP must be inserted last. This is required because
+	// we maintain a separate index for the most recent TWAP records that
```
Comment (p0mvn) [inline x/twap/keeper.go:56]:
```suggestion
	// we maintain a separate index for the most recent records.
```

File: x/twap/keeper.go [hunk 5]
```diff
@@ -40,3 +42,39 @@ func (k Keeper) SetParams(ctx sdk.Context, params types.Params) {
...
+	// is updated by storing new records.
```
Comment (p0mvn) [inline x/twap/keeper.go:57]:
```suggestion
	// It is updated by storing new records.
```

File: x/twap/keeper.go [hunk 6]
```diff
@@ -40,3 +42,39 @@ func (k Keeper) SetParams(ctx sdk.Context, params types.Params) {
...
+	})
+
+	for _, twap := range genState.Twaps {
+		k.storeNewRecord(ctx, twap)
+	}
+}
+
+// ExportGenesis returns the twap module's exported genesis.
+func (k Keeper) ExportGenesis(ctx sdk.Context) *types.GenesisState {
+	// These are ordered in increasing order, guaranteed but the iterator
```
Comment (p0mvn) [inline x/twap/keeper.go:69]:
```suggestion
	// These are ordered in increasing order, guaranteed by the iterator
```

File: x/twap/types/genesis.go [hunk 7]
```diff
@@ -5,7 +5,7 @@ import (
 	"fmt"
 )
 
-// NewGenesisState returns genesis state with the given parameters and twap records.
+// NewGenesisState reeturns genesis state with the given parameters and twap records.
```
Comment (p0mvn) [inline x/twap/types/genesis.go:8]:
```suggestion
// NewGenesisState returns genesis state with the given parameters and twap records.
```

File: x/twap/types/export_test.go [hunk 8]
```diff
@@ -1,5 +1,6 @@
 package types
 
+// validate validates the twap record, returns nil on success, error otherwise.
```
Comment (p0mvn) [inline x/twap/types/export_test.go:3]:
```suggestion
```

**Comment by p0mvn** [comment]
R4R

**Review by ValarDragon — APPROVED** [review]
Great job on the tests

**Pull request merged**